processor/src/main/java/com/linecorp/decaton/processor/metrics/Metrics.java

```diff
@@ -79,14 +76,21 @@ protected AbstractMetrics() {
         @Override
         public void close() {
             synchronized (meterRefCounts) {
-                for (Meter meter : meters) {
-                    Meter.Id id = meter.getId();
-                    int count = meterRefCounts.get(id).decrementAndGet();
-                    if (count == 0) {
+                // traverse from the end to avoid arrayCopy
+                for (ListIterator<Meter> iterator = meters.listIterator(meters.size() - 1); iterator.hasNext(); ) {
+                    Meter meter = iterator.previous();
+                    Id id = meter.getId();
+                    AtomicInteger count = meterRefCounts.get(id);
+                    if (count == null) {
+                        throw new IllegalStateException("Missing reference to meter: " + id);
```

Comment:
> I don't think so, races can still happen since this static map is shared by several AbstractMetrics instances

hm, can it?
I thought current synchronize makes sure atomic 1. refcount decrement, 2. `meterRefCounts` entry drop and 3. `iterator.remove();` without possibility of partial failure in the middle so we never expect to see null here.

Reply:
Yes, my comment is deprecated. Sorry.
Hence throwing here meaning "wtf it's impossible"